add support for optional chaining

sanex3339 opened the pull request:
Spec:
https://github.com/estree/estree/pull/204/files

sanex3339 commented:
https://github.com/acornjs/acorn/pull/891 merged, will back to this PR this weekend

sanex3339 commented:
Updated this PR

sanex3339 commented:
@michaelficarra :eyes:

michaelficarra commented on test/compare-acorn-es2020/optional-chaining.expected.js:
```diff
@@ -0,0 +1,16 @@
+obj?.aaa?.bbb;
+obj.aaa.bbb;
+obj.aaa?.bbb;
+obj?.aaa.bbb;
```
This should be `(obj?.aaa).bbb`. I don't trust these expectations. Can you add an assertion to the tests that the result of parsing the output of the codegen is the same as the result of parsing the original input?

sanex3339 replied:
You are right. But i'm not good with parenthesis. Anybody can help?

sanex3339 replied:
@michaelficarra updated. Now expected results same as `babel` results.
I updated logic, but not sure that did it right, but all test are passed now.

michaelficarra replied:
@sanex3339 You've not added the round-tripping assertion I asked for.

sanex3339 replied:
@michaelficarra in which test file i can add it?

michaelficarra replied:
I would put it right [here](https://github.com/estools/escodegen/pull/412/files#diff-cacb5c391de59a37897c31fde28d1783R49). And the same for `testMin`.

sanex3339 replied:
Hmm. Looks like AST tree of the output cannot be 1 to 1 matched with input AST tree, because of the parens and how AST tree will be generated.

But the generated code is the same as generated code from the babel. 
Tried it here:
https://babeljs.io/repl/#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=PYIwVgdAhjEvBuAUKSMoH46JeL7sRlV8Yt4ikAKVaGASkORvDqnvJxbFPaepJsO_Kt16MKEAMYzGAEwXMxQwlhlT5igXiGdKqANroAugYrHi4IzGNYz8C7h4RrUU-cvPXtlx-1hventHbkC7P1CTYXsQEMFAmLidMN8HJUMoxIN1YyCFORDlBPM7HLyFRyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.10.2&externalPlugins=babel-plugin-minify-dead-code-elimination%400.5.1

sanex3339 replied:
that is wat i'm talking about:
![image](https://user-images.githubusercontent.com/9210534/84543317-d5b2e680-ad03-11ea-8342-6575b639416b.png)

sanex3339 replied:
@michaelficarra beat it! Added AST comparison tests. Now AST is the same, but the output code is different from Babel output, because we have to keep some parenthesis to make input AST the same as output AST

sanex3339 commented:
@michaelficarra should we support node@4 in travis? I added `chai-exclude` package to deep compare  AST-trees without `start` and `end` properties, but looks like it doesn't support node@4.

I can remove node@4 from travis.yml and add node@10 instead

sanex3339 commented:
Ok. I removed `node@4` from travis.yml file and added `node@10` instead.
Also, I updated `engines` section with `>=6.0` value.

Node 4 is few years as deprecated, so I think it's the time to forget its.

sanex3339 commented:
@michaelficarra also i created PR to the `estraverse` to support `ChainExpression` node. 
https://github.com/estools/estraverse/pull/113

So it is best to merge `estravserse` PR first and after that, I'll update this PR with the use of `Syntax.ChainExpression`.

sanex3339 commented:
Any news?

sanex3339 commented:
Any ETA when this PR can be reviewed?

dilame commented:
Hi guys! This PR is very important for me and many people in the world! Please, merge it

mikhailrojo commented:
I upvote for fast merge

sanex3339 commented:
@Constellation can you check and merge this?

sanex3339 commented:
@michaelficarra re-ping

sanex3339 commented:
@michaelficarra re-ping

dilame commented:
i support this re-ping

sanex3339 commented:
@michaelficarra please, re-check this

sanex3339 commented:
@michaelficarra re-ping

dilame commented:
@michaelficarra re-ping

sanex3339 commented:
@michaelficarra ping

sanex3339 commented:
![image](https://user-images.githubusercontent.com/9210534/87241265-e8672b00-c429-11ea-92c1-22b1ba753319.png)
Still waiting...

sanex3339 commented:
@michaelficarra ping

sanex3339 commented:
@michaelficarra please check this. More than a month I'm waiting for your approval.

dilame commented:
@michaelficarra it is insulting the feelings of contributors!

dilame commented:
@michaelficarra !!!

sanex3339 commented:
I hope if this pr merged, it takes much less time to merge the PR with Nullish Coalescing support.
https://github.com/estools/escodegen/pull/417

sanex3339 commented:
@michaelficarra ping

sanex3339 commented:
@michaelficarra i really think that if you spend 5 minutes to re-check this PR with ALL your requested changes and merge it - all users who are using this package will happy.

It is absurd that you haven't 5 minutes during 2 months to re-check this.

papandreou commented:
@sanex3339, why don't you create a fork of escodegen and publish your changes there? That way you won't be blocked on getting these PRs merged.

sanex3339 commented:
The fork is the last option in cases like that. I'll still try to avoid it.

papandreou commented:
Why try to avoid it? It’s a fairly easy thing to do even temporarily. It’s also very liberating to not be dependent on others.

sanex3339 commented:
@michaelficarra ping

dilame commented:
@sanex3339 it's true - it is very easy to fork a project. 
I guess you think its hard because you think you need to publish it in NPM, but in fact, you can declare your GitHub fork URL as a dependency in `package.json`. i know how to make it right. If you want you can contact me in telegram `@bowzee` - I'll help you with that.

bakkot commented:
@sanex3339 tests appear not to cover optional call (`a?.()`)

bakkot commented on escodegen.js:
```diff
@@ -1943,10 +1966,14 @@
                     }
                 }
-                result.push('.');
+                result.push(expr.optional ? '?.' : '.');
```
The additional ` ` added by the preceding block is unnecessary when using optional chaining - `1?.a` works fine. This looks like it will codegen that as `1 ?.a`.

sanex3339 replied:
Nice catch!

sanex3339 replied:
https://github.com/estools/escodegen/pull/412/commits/ba05f2050732d9f9504237f708e180ba9320d4b2

sanex3339 commented:
> @sanex3339 tests appear not to cover optional call (`a?.()`)
https://github.com/estools/escodegen/pull/412/commits/f23abd632b16c610952af768d3060ab229309cac

sanex3339 commented:
Now looks like it's ok

bakkot commented:
@sanex3339 It seems like there ought to be tests that mix optional calls with member access  - that is, `(a?.b)()`, `a?.b()`, `a?.().b`, `(a?.()).b`, `(a?.b.c)()`, `a?.b.c()`, `a.b?.().c`, `(a.b?.()).c`, etc.

sanex3339 commented:
@bakkot Added more tests and fixed logic. Not sure that i did it right, but all tests are passed.

sanex3339 commented:
Simplified logic

bakkot commented:
With regards to the actual implementation, it seems like it would be better to introduce a new precedence level for optional chains just below `Call`, and have `ChainExpression` parenthesize its result at that precedence - something along the lines of

```js
Precedence = {
        [...]
        Postfix: 15,
        OptionalChain: 16,
        Call: 17,
        New: 18,
        TaggedTemplate: 19,
        Member: 20,
        Primary: 21
}

[...]

        ChainExpression: function (expr, precedence, flags) {
            var expression = expr.expression;

            var result = this[expression.type](expression, precedence, flags);
            return parenthesize(result, Precedence.OptionalChain, precedence);
        },
```

Then you wouldn't need to mess around with `precedence` in `MemberExpression` and `CallExpression`. (I haven't tested this, but it seems like it should work.)

sanex3339 commented:
> With regards to the actual implementation, it seems like it would be better to introduce a new precedence level for optional chains just below `Call`, and have `ChainExpression` parenthesize its result at that precedence - something along the lines of
> 
> ```js
> Precedence = {
>         [...]
>         Postfix: 15,
>         OptionalChain: 16,
>         Call: 17,
>         New: 18,
>         TaggedTemplate: 19,
>         Member: 20,
>         Primary: 21
> }
> 
> [...]
> 
>         ChainExpression: function (expr, precedence, flags) {
>             var result = this[expression.type](expr.expression, precedence, flags);
>             return parenthesize(result, Precedence.OptionalChain, precedence);
>         },
> ```
> 
> Then you wouldn't need to mess around with `precedence` in `MemberExpression` and `CallExpression`. (I haven't tested this, but it seems like it should work.)

https://github.com/estools/escodegen/pull/412/commits/ac6361852f61f1fd3e0c16e3987c26f6272e396f

bakkot commented:
Some last things:

- I'm pretty sure you should be using `this.generateExpression` rather than `this[expression.type]`, just based on the rest of the file
- The precedence to generate the things in the chain should be `Precedence.OptionalChain`, not `precedence` (sorry, I wrote the wrong thing in the previous comment)
- In `ChainExpression`, when you're going to add parentheses (i.e. `Precedence.OptionalChaining < precedence`) you should set the `F_ALLOW_CALL` bit

The last couple of points are necessary so that `new (a?.b().c)` generates correctly and `new (a?.b())` generates without additional parentheses and so on; you should add some tests for these and similar cases.

Putting that together, I think you want

```js
ChainExpression: function (expr, precedence, flags) {
    if (Precedence.OptionalChaining < precedence) {
        flags |= F_ALLOW_CALL;
    }

    var result = this.generateExpression(expr.expression, Precedence.OptionalChaining, flags);
    return parenthesize(result, Precedence.OptionalChaining, precedence);
},
```

sanex3339 commented:
Thank you, i'll fix ot soon. Are you new maintainer of this repository? If no, it would be nice if you are.

bakkot commented:
No, just a passer-by.

sanex3339 commented:
@bakkot about `new` tests.
`(a?.b().c)` part of `new (a?.b().c)` expression is just a `callee`, without arguments. So escodegen WILL add empty brackets to it:
`new (a?.b().c)()`;
So, parenthesis is necessary here, because optional chaining as part of `new` expressions is not supported by spec and AST parsers: https://github.com/tc39/proposal-optional-chaining#not-supported

sanex3339 commented:
Added more tests with `new`

sanex3339 commented:
@michaelficarra may you check it now, please

bakkot commented:
>  escodegen WILL add empty brackets to it:

Only when the `parentheses` option is set. They're not required, as you can see by the fact that they're not included in your `minimal` test.

> So, parenthesis is necessary here

The parentheses around the `(a?.b().c)` are necessary, but the trailing `()` parentheses are not.

sanex3339 commented:
@michaelficarra we're still here. Ping.

bakkot commented:
@sanex3339 I strongly suspect that pinging the maintainers repeatedly is not going to get this merged faster.

michaelficarra commented:
I will happily merge and release as soon as @bakkot is done with his review. Also, what's the reason for dropping node 4 support and upgrading dependencies? It seems unrelated to this PR.

sanex3339 commented:
> I will happily merge and release as soon as @bakkot is done with his review. Also, what's the reason for dropping node 4 support and upgrading dependencies? It seems unrelated to this PR.

I added chai-exclude package to deep compare AST-trees without start and end properties, but looks like it doesn't support node@4.

sanex3339 commented:
@bakkot are there any additional changes that you want?

bakkot commented:
@michaelficarra the implementation itself LGTM but I don't feel qualified to review the change to supported Node versions and the test harness.

michaelficarra commented:
@sanex3339 Can you extract the node/dependency changes to a separate PR? I agree that dropping node 4 support is reasonable.

sanex3339 commented:
@michaelficarra `chai-exclude` package requires node@6.
You asked me to add AST trees comparison, and `chai-exclude` is the easy way to remove `start` and `end` properties for AST comparison

sanex3339 commented on test/compare-acorn-es2020.js:
```diff
@@ -0,0 +1,95 @@
+    actualTree = acorn.parse(actual, options);
+
+    expect(actual).to.be.equal(expected);
+    expect(tree).excludingEvery(['start', 'end']).to.deep.equal(actualTree);
```
I'm talking about this place

sanex3339 commented:
> I can extract AST-tree comparison and node version update into another PR. But it's a strange change because AST comparison helps me much with this PR.

@michaelficarra so, I need your decision.

michaelficarra commented:
@sanex3339 All I am suggesting is that we drop node 4 support in its own PR. I'll merge that, then you can rebase this PR on top of that one. You can leave the dependency changes in this one, since I see now that they're all related to this PR.

sanex3339 commented:
Lets keep it here

sanex3339 commented:
Ok, i'll create pr with node version update

sanex3339 commented:
PR with node update https://github.com/estools/escodegen/pull/418

michaelficarra commented:
@sanex3339 Please rebase.

sanex3339 commented:
> @sanex3339 Please rebase.

@michaelficarra done

michaelficarra commented on test/compare-acorn-es2020/optional-chaining.expected.js:
```diff
@@ -0,0 +1,47 @@
+(a.b?.()).c;
+(a.b?.())?.c;
+new (a?.b().c)();
+new (a?.b())();
```
The trailing parens on these two lines are unnecessary. I think we're missing some special handling in the generation of `NewExpression`.

michaelficarra commented on test/compare-acorn-es2020/optional-chaining.expected.js:
```diff
@@ -0,0 +1,47 @@
+new (a?.b().c)();
+new (a?.b())();
+new (a?.b().c)();
+new (a?.b())();
```
The trailing parens on these four lines are unnecessary. I think we're missing some special handling in the generation of `NewExpression`.

sanex3339 replied:
https://github.com/estools/escodegen/pull/412/commits/8f05fb08753c49aa1a0cc6e14c593694e3b06db8

sanex3339 replied:
If you want, i can change early return with condition like:
```
        if (
                !(expr.callee.type === Syntax.ChainExpression && !length)
                && (
                    !(flags & F_ALLOW_UNPARATH_NEW)
                    || parentheses
                    || length > 0
                )
            )
```

bakkot replied:
@michaelficarra Those parentheses were included only because the `parentheses` option [defaults to true](https://github.com/estools/escodegen/blob/eee573bee25bcc42ef405e685d6930eda62e33de/escodegen.js#L191). They were not correctly omitted in the `min.js` test.

@sanex3339 You should revert that commit.

sanex3339 replied:
Ok

sanex3339 replied:
https://github.com/estools/escodegen/pull/412/commits/eb9cbb0bfd1ea745c1a12632f66b31f602573ded

michaelficarra replied:
Oh, GitHub collapsed the `.min.js` diff and I didn't even notice those tests. Fixed: https://github.com/estools/escodegen/commit/e8e2e81675ce74a6f538ad0274fd72ada26fb17d

michaelficarra commented:
Published v2.0.0 (since I got yelled at last time I dropped support for an ancient node version in a minor release).

sanex3339 commented:
Thank you!
